llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -2637,18 +2637,15 @@ Instruction *InstCombinerImpl::visitAnd(BinaryOperator &I) {
   // This is a generous interpretation for noimplicitfloat, this is not a true
   // floating-point operation.
   //
-  // Assumes any IEEE-represented type has the sign bit in the high bit.
+  // Assumes any floating point type has the sign bit in the high bit.
   // TODO: Unify with APInt matcher. This version allows undef unlike m_APInt
   Value *CastOp;
   if (match(Op0, m_ElementWiseBitCast(m_Value(CastOp))) &&
-      match(Op1, m_MaxSignedValue()) &&
+      CastOp->getType()->isFPOrFPVectorTy() && match(Op1, m_MaxSignedValue()) &&
```

Comment:
It might be worth wrapping this in a predicate with this explicit property

Reply:
How about adding new matchers like `m_IntValue/m_FPValue`? They are very useful to work with the bitcast matcher.

Reply:
@arsenm Do you mean the property "has sign bit in high bit"?

Reply:
Yes